### DISCUSSION TITLE
Testing against only specific opponent

### DISCUSSION THREAD

**Discussion by rnz1234** [yasserfarouk/scml#46] [one-shot-qa] [answered]
I want to test my agent against only one specific agent, to analyze it's decision making and optimize it.
I have the following problems:
1. I add only my agent and the opponent through competitors, but the simulation add other agents as well (Random agent, Ran@)
2. I see that the simulation randomly doesn't give my agent to start working, probably giving it no contracts. I want a quite deterministic setting that I can work on to optimize my agent and just can't get to it

is there a way to get to a deterministic 1 on 1 kind of setting ?

thanks

**Comment by yasserfarouk** [comment, marked as answer]
You can control all aspects of world generation if you directly call `generate()` method or `SCML2020OneshotWorld` constructor.

> I add only my agent and the opponent through competitors, but the simulation add other agents as well (Random agent, Ran@)

The system will fill extra factories with built-in agents. You need to have exactly the same number of profiles and agent types specified to control this. You can find examples of how to do so [here](http://yasserm.com/scml/scml2020docs/tutorials/04.experiments.html)

> I see that the simulation randomly doesn't give my agent to start working, probably giving it no contracts. I want a quite deterministic setting that I can work on to optimize my agent and just can't get to it

The same link above shows how you can do exactly that.

**Reply by yasserfarouk** [reply]
Hi @rnz1234 
Does the above example answer your question?
